Create inline_link_in_paragraph_is_distinguishable_ be4d0c.md (#1010)
act-rules/act-rules.github.io: _rules/inline_link_in_paragraph_is_distinguishable_be4d0c.md

---
id: be4d0c
name: Inline link in paragraph is distinguishable
rule_type: atomic

description: |
  This rule checks that links that are embedded in a paragraph have a way to distinguish them as links compared to the rest of the text

accessibility_requirements: # Remove whatever is not applicable
  wcag20:1.4.1: # Use of Color (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling

> acknowledgements:

Reviewer: Would you mind listing me as a previous author:
https://github.com/act-rules/act-rules.github.io/blob/develop/__deprecated__/_drafts/outdated/SC1-4-1-link-style-in-text.md

  authors:
    - Brian Bors
    - Carlos Duarte
---

## Applicability

This rule applies to any HTML element which:

- has the [semantic role][] of ['link'][link] or a [semantic role][] that inherits from the ['link'][link] role; and

> - is a [descendant][] of a `p` element; and

Reviewer 2: Should the `<p>` element be an HTML element?

> - is a [descendant][] of a `p` element; and

Author: The initial version used "block of text" which we found to be ambiguous. So we opted to focus this rule on paragraphs (it's even in the name). When we get this done, we can then move to another "blocks of text".

> - is a [descendant][] of a `p` element; and

Author: We're including `li` and `td` but leaving out `div` because you can have all content of a page being a descendant of the same `div`.

- has [descendant][] [text nodes][text node] that are [visible][]; and
- the `p` element has other [descendant][] [visible][] [text nodes][text node].

## Expectation

Each target element has either:

> - An icon immediately before, after or in the link denoting it as a link

Reviewer 2: What is an "icon"?

> - An icon immediately before, after or in the link denoting it as a link

Reviewer: Not the change in order, so someone can't read "immediately ... in the link" in there.

- Text immediately before, after or in the link denoting it as a link

> - a [distinguishing style][] compared to the other text not based on color alone

Reviewer: The distinguishing style definition compares two elements, here you are comparing the link element to text nodes. I don't think that works. Do you mean to compare the link element to the parent element of the text node? If so, there are a bunch of problems with the distinguishing style definition that needs to be worked through. For example:

```html
<style>
  * { box-shadow: #F00 0 0 5px; color: black; }
  a { text-decoration: none; }
</style>
<p>Lorem ipsum <a href="">dolar sit amet</a> consectetur.</p>
```

This link is still distinguishable, because it's shadow is an inline one, whereas the shadow of the rest of the text is on the block.

Author: Updated the expectation (and assumptions) for 

- comparing elements instead of element and text nodes
- usage of border and box-shadow as distinguishing factors

> - a different color compared to the rest of the text in the paragraph that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) difference with the rest of the text and a [distinguishing style][] both when the target element [gains focus][gain focus] and the target element is [hovered][]

Reviewer 2: I'm assuming that by "the rest of the text in the paragraph", the rule is talking about "the other descendant visible text nodes" of the ancestor `<p>` element?

> - a different color compared to the rest of the text in the paragraph that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) difference with the rest of the text and a [distinguishing style][] both when the target element [gains focus][gain focus] and the target element is [hovered][]

Reviewer: This will need to consider differences in background color as well. Either the text color, or the background color needs to have sufficient contrast with that of the surrounding text.

Author: Added to the expectation

> - a different color compared to the rest of the text in the paragraph that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) difference with the rest of the text and a [distinguishing style][] both when the target element [gains focus][gain focus] and the target element is [hovered][]

Author: That is correct. If the rest of the text has more than one color then color can't be a distinguishing factor. Need to add that assumption to the text of the expectation.

> - a different color compared to the rest of the text in the paragraph that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) difference with the rest of the text and a [distinguishing style][] both when the target element [gains focus][gain focus] and the target element is [hovered][]

Author: Even if the text if all of the same color?

## Assumptions

> This rule assumes that the link is distinguishable from the rest of the text with color, which means it fails SC 1.4.1 when there is not another way to distinguish it.

Reviewer: I don't understand this assumption. The expectation tests for color.

Reviewer 3: The expectation is indeed testing for color, but it needs to be a contrast ratio of at least 3:1 for the expectation to be met.

If we leave this assumption out of it; there could be a scenario where the link is not(!) a different color from the rest of the text and this rule would fail it even though it should not fail the SC. (Because you should only fail this SC if you communicate something with color without communicating it in a different way.

Does that help you understand this assumption?

Reviewer 4: indeed, if the link is the same color as the surrounding text, this rule is not applicable (it's an odd gap in WCAG since in those cases it's deemed that the link is indistinguishable for all sighted users, not just those with color deficiency/vision impairment that requires at least a 3:1 contrast - though paradoxically keyboard/AT users are advantages because THEY can get to the links without any problem)

> This rule assumes that the link is distinguishable from the rest of the text with color, which means it fails SC 1.4.1 when there is not another way to distinguish it.

Reviewer: I think these assumptions are relevant to this rule as well:
https://github.com/act-rules/act-rules.github.io/blob/develop/__deprecated__/_drafts/outdated/SC1-4-1-link-style-in-text.md#assumptions

## Accessibility Support

_No accessibility support issues known._

## Background

- [Understanding Success Criterion 1.4.1: Use of Color](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html)
- [G183: Using a contrast ratio of 3:1 with surrounding text and providing additional visual cues on focus for links or controls where color alone is used to identify them](https://www.w3.org/WAI/WCAG21/Techniques/general/G183)
- [F73: Failure of Success Criterion 1.4.1 due to creating links that are not visually evident without color vision](https://www.w3.org/WAI/WCAG21/Techniques/failures/F73)
- [C15: Using CSS to change the presentation of a user interface component when it receives focus](https://www.w3.org/WAI/WCAG21/Techniques/css/C15)

## Test Cases

### Passed

#### Passed Example 1

This is a link that is a descendant of a paragraph element and it uses the default styling of links which makes it underlined in most browsers which is a distinguish style.

```html

> <p>Read about WAI on the <a href="http://w3.org/WAI">WAI webpage</a>.</p>

Reviewer: I'm not sure it's a good idea to rely on default styles for this. You should probably define the styles you want. People with their machine set up in unusual ways can cause this page to be a failure.

```

#### Passed Example 2

This is a link that is a descendant of a paragraph element and the underline is removed but an icon is added to denote it as a link.

```html
<html>
	<head>

> 		<title>Passed example 2 for rule be4d0c</title>

Reviewer: Why did you include titles in these test cases? This is inconsistent with what we've done for other rules. I think these should be removed.

Reviewer 3: Because these test cases have head elements. Most test cases don't have head elements, but if they do they should also have title elements, otherwise they would fail SC 2.4.2 right?

```html
<html>
	<head>
		<title>Passed example 3 for rule be4d0c</title>
		<style>
			a.test {
				text-decoration: none;
			}
		</style>
	</head>
	<body>
		<p>Read about WAI on the <a class="test" href="http://w3.org/WAI">WAI webpage by following this link</a>.</p>
	</body>
</html>
```

#### Passed Example 4

This is a link that is a descendant of a paragraph element and the underline is removed but the link has a distinguishing box-shadow which is a distinguishing style.

```html
<html>
	<head>
		<title>Passed example 4 for rule be4d0c</title>
		<style>
			a.test {
				text-decoration: none;
				box-shadow: 4px 4px;
			}
		</style>
	</head>
	<body>
		<p>Read about WAI on the <a class="test" href="http://w3.org/WAI">WAI webpage</a>.</p>
	</body>
</html>
```

#### Passed Example 5

This is a link that is a descendant of a paragraph element and the underline is removed but the link has a text contrast of more than 3:1 compared to the other text in the paragraph and when it receives focus or hover an underline appears.

```html
<html>
	<head>
		<title>Passed example 5 for rule be4d0c</title>
		<style>
			a {
				text-decoration: none;
				color: #d14826;
			}

> 			a:hover {

Reviewer: ```suggestion
			a:hover, a:focus {
```

				text-decoration: underline;
			}
			a:focus {
				text-decoration: underline;
			}
		</style>
	</head>
	<body>
		<p>Read about WAI on the <a href="http://w3.org/WAI">WAI webpage</a>.</p>
	</body>
</html>
```

### Failed

#### Failed Example 1

This is a link that is a descendant of a paragraph element and the underline is removed but the link has no other way of being recognized as a link.

```html
<html>
	<head>
		<title>Failed example 1 for rule be4d0c</title>
		<style>
			a.test {
				text-decoration: none;
			}
		</style>
	</head>
	<body>
		<p>Read about WAI on the <a class="test" href="http://w3.org/WAI">WAI webpage</a>.</p>
	</body>
</html>
```

> ### Inapplicable

Reviewer: Inapplicable example 3 - 6 need to have underline removed for them to be a meaningful test.

Reviewer 3: Do you mean to say that inapplicable examples should fail the SC if they would where to be applicable?

And if so, why only 3-6, why not also add this to 1 and 2?

Reviewer 5: I do not see the need to remove the underline from examples 3 - 6. By keeping the underline, the test cases focus on the relevant parts of the (in)applicability.

I would suggest adding an inapplicable test case with a link that is not distunguised using colour. I see that this is handled as an assumption and may not need its own test case?

#### Inapplicable Example 1

This is no semantic role of link within this example.

```html
<p>Read about WAI on the <u>underlined text</u>.</p>
```

#### Inapplicable Example 2

This is a link that is not a descendant of a paragraph element.

```html
Read about WAI on the <a href="http://w3.org/WAI">WAI webpage</a>.
```

#### Inapplicable Example 3

This link has no descendant text nodes.

```html

> <p>Read about WAI on the <a href="http://w3.org/WAI"></a>.</p>

Reviewer: You should maybe put an image in here, and remove the underline from the link, to indicate that this would fail if it did have text. Otherwise it's not much of a test IMO.

> ---

Reviewer: Can you change the file name to use `-` instead of `_`? That's what we've used in all other rules.

id: be4d0c

> name: Inline link is distinguishable

Reviewer: The file name mentions "in_paragraph". Can you make this consistent?

rule_type: atomic

description: |

>   This rule checks that links that are embedded in a paragraph, list or cell have a way to distinguish them as links compared to the rest of the text

Reviewer: ```suggestion
  This rule checks that links that are embedded in a paragraph, list or cell have a way to distinguish them from the surrounding text
```

accessibility_requirements: # Remove whatever is not applicable
  wcag20:1.4.1: # Use of Color (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling

acknowledgements:
  authors:
    - Brian Bors
    - Carlos Duarte
  previous authors:
    - Wilco Fiers
---

## Applicability

This rule applies to any HTML element which:

- has the [semantic role][] of ['link'][link] or a [semantic role][] that inherits from the ['link'][link] role; and

> - is a [descendant][] of a `p`, `li` or `td` HTML elements (henceforth referred to as the _ancestor_ element) in the [flat tree][]; and

Reviewer: What's the reason for this scope? Why just these three elements, and not headings for instance, or a div with role=listitem? This should be either explained in a note, or changed.

Author: The rule looks at links within paragraphs. We agreed that besides the `p` element, text within list items and table cells would be considered paragraphs also. But a heading should not be a paragraph. 
I've extended the applicability to consider lisitem and cell roles

- has [descendant][] [text nodes][text node] in the [flat tree][] that are [visible][]; and

> - the _ancestor_ element has other [descendant][] [visible][] [text nodes][text node] in the [flat tree][].

Reviewer: I don't think something like this should be applicable:

```html
<p>
  Hello world
  <br><br>
  <a href="">Next page</a>
</p>
```

Author: Agree. But the `<br>` element should not be used like that according to the [HTML spec](https://html.spec.whatwg.org/#the-br-element). The spec says:
- `br` elements must be used only for line breaks that are actually part of the content, as in poems or addresses.
- `br` elements must not be used for separating thematic groups in a paragraph (which is what you are doing in that example)

I don't think we consider incorrect uses of HTML elements in other rules, so I don't think we need to do that in this rule, and I won't change the applicability based on this (unless there are other arguments).

> ## Expectation

Reviewer: This is really very difficult to read. I question whether or not we should try to do this in a single rule. It seems to me a better strategy here would have been to create a few rules, each checking a different way to make a link distinguishable, and have a composite rule that aggregates them all.

The way this is formulated now, I don't think it follows the "atomic" requirement from the ACT rules format.

Author: Good point. I'll make this a composite.

> Each target element has either:

Reviewer: ```suggestion
Each target element has one of the following:
```

> - content (such as an image or text) visually located inside, or immediately before or after the test target, that identifies it as a link; or

Reviewer: I don't like the word "identifies" here. Identifies has a specific meaning in WCAG. I think a softer word is better hee:

```suggestion
- content (such as an image or text) visually located inside, or immediately before or after the test target, that indicates the test target is as a link; or
```

> - a [distinguishing style][] not based on color alone, compared to the _ancestor_ element and to any of its [descendant][] HTML elements that have [visible][] [text nodes][text node]; or

Reviewer: > the ancestor element

Which ancestor element? The one from the applicability? What if there is more than one?

Also, the way this is worded, if the paragraph has two links with the same style, this would not be true and those things could fail the rule.

Author: Nice catch (both)

> - a [distinguishing style][] not based on color alone, compared to the _ancestor_ element and to any of its [descendant][] HTML elements that have [visible][] [text nodes][text node]; or

Reviewer: I don't think "a different X, compared to Y" is correct English.

- a `box-shadow` style property with a computed [color value](https://drafts.csswg.org/css-backgrounds/#shadow-color) different from the `background-color` of the target element; or

> - at least one of the `border-top-color`, `border-right-color`, `border-bottom-color` and `border-left-color` style properties with a computed [color value](https://drafts.csswg.org/css-backgrounds/#border-color) different from the `background-color` of the target element; or

Reviewer: You'll need to check the width of the border here. If it has a 0px with, this should fail.

- This rule assumes that if `box-shadow` is used in the different visible text nodes of the _ancestor_ element then `box-shadow` can not be a distinguishing factor.
- This rule assumes that if `border` is used in the different visible text nodes of the _ancestor_ element then `border` can not be a distinguishing factor.

## Accessibility Support

_No accessibility support issues known._

## Background

- [Understanding Success Criterion 1.4.1: Use of Color](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html)
- [G183: Using a contrast ratio of 3:1 with surrounding text and providing additional visual cues on focus for links or controls where color alone is used to identify them](https://www.w3.org/WAI/WCAG21/Techniques/general/G183)
- [F73: Failure of Success Criterion 1.4.1 due to creating links that are not visually evident without color vision](https://www.w3.org/WAI/WCAG21/Techniques/failures/F73)
- [C15: Using CSS to change the presentation of a user interface component when it receives focus](https://www.w3.org/WAI/WCAG21/Techniques/css/C15)

> ## Test Cases

Reviewer: There are no examples for things that inherit from link. I think only DPUB-ARIA has things that inherit from link, so use one of those.

Author: I've added an example with an element with the `doc-glossref` role

---
id: be4d0c
name: Inline link is distinguishable
rule_type: atomic

description: |
  This rule checks that links that are embedded in a paragraph, list or cell have a way to distinguish them as links compared to the rest of the text

accessibility_requirements: # Remove whatever is not applicable
  wcag20:1.4.1: # Use of Color (A)
    forConformance: true
    failed: not satisfied
    passed: further testing needed
    inapplicable: further testing needed

input_aspects:
  - DOM Tree
  - CSS Styling

acknowledgements:
  authors:
    - Brian Bors
	- Carlos Duarte
  previous authors:
    - Wilco Fiers
---

## Applicability

This rule applies to any HTML element which:

- has the [semantic role][] of ['link'][link] or a [semantic role][] that inherits from the ['link'][link] role; and
- is a [descendant][] of a `p`, `li` or `td` HTML elements (henceforth referred to as the _ancestor_ element); and
- has [descendant][] [text nodes][text node] that are [visible][]; and

> - the _ancestor_ element has other [descendant][] [visible][] [text nodes][text node].

Author: Updated applicability to look at flat tree

## Expectation

Each target element has either:

> - content (such as an image or text) inside, or immediately before or after the test target, that identifies it as a link; or

Author: What exactly do you think it's ambiguous here?

> - content (such as an image or text) inside, or immediately before or after the test target, that identifies it as a link; or

Reviewer 2: > content (such as an image or text) inside [...]

"Content" by itself is pretty broad and open to interpretation. Given that the rule explicitly refers to "content [...] that identifies [the link] as a link", on second thought I don't think that's an issue though.

> [...] immediately before or after the test target [...].

In visual order or in tree order?

- a [distinguishing style][] not based on color alone, compared to the _ancestor_ element and to any of its [descendant][] HTML elements that have [visible][] [text nodes][text node]; or

> - a `box-shadow` style property with a color value different from `transparent` or the `background-color` of the target element; or

Author: Sorted this by using the color definitions from #833 following your suggestion

- a `border` style property with a color value different from `transparent` or the `background-color` of the target element; or
- a different color, compared to the color of the other descendant visible text nodes of the same _ancestor_ element (if all have the same color), that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) and a [distinguishing style][] both when the target element [gains focus][focused] and the target element is [hovered][]; or

> - a different background color, compared to the background color of the other descendant visible text nodes of the same _ancestor_ element (if all have the same background color), that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) and a [distinguishing style][] both when the target element [gains focus][focused] and the target element is [hovered][].

Reviewer 2: The expectation is currently unclear on at least two important points:

- How to deal with short-/longhands? We inconsistently refer to both a shorthand (`border`) and a longhand (`box-shadow`) property. I'm thinking we'd want to only ever refer to longhands as these are what conformant implementations of CSS must expand to anyway (https://drafts.csswg.org/css-cascade/#shorthand and https://drafts.csswg.org/cssom/#concept-declarations-specified-order).

- How to deal with semi-transparent border and box shadow colors in which case alpha compositing comes into play? One of the points about the definitions of "foreground color" and "background color" as defined by #833 was to avoid having to explicitly deal with alpha compositing at the rule level. If we refer to specific CSS properties, there's not really a way around this as I see it though.

> - a different background color, compared to the background color of the other descendant visible text nodes of the same _ancestor_ element (if all have the same background color), that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) and a [distinguishing style][] both when the target element [gains focus][focused] and the target element is [hovered][].

Author: The (foreground and background) color definitions in #833 might return multiple colors. Initially, I thought that would be a problem for this rule. But given that there is an assumption stating that if multiple colors are in use then color can't be a distinguishing factor then that problem goes away and I believe I can rewrite the expectation using #833 definitions.

> - a different background color, compared to the background color of the other descendant visible text nodes of the same _ancestor_ element (if all have the same background color), that has at least a 3:1 [contrast ratio](https://www.w3.org/TR/WCAG21/#dfn-contrast-ratio) and a [distinguishing style][] both when the target element [gains focus][focused] and the target element is [hovered][].

Author: Agreed with using only longhand properties. Replaced `border-color` by the corresponding longhand properties.

	</body>
</html>
```

#### Passed Example 4

This is a link that is a descendant of a paragraph element and the underline is removed but the link has a distinguishing box-shadow.

```html
<html>
	<head>
		<title>Passed example 4 for rule be4d0c</title>
		<style>
			* {
				text-decoration: none;
			}
			a.test {
				text-decoration: none;
				box-shadow: 4px 4px;
			}
		</style>
	</head>
	<body>
		<p>Read about WAI on the <a class="test" href="http://w3.org/WAI">WAI webpage</a>.</p>
	</body>
</html>
```

#### Passed Example 5

This is a link that is a descendant of a paragraph element and the underline is removed but the link has a text contrast of more than 3:1 compared to the other text in the paragraph and when it receives focus or hover an underline appears.

```html
<html>
	<head>
		<title>Passed example 5 for rule be4d0c</title>
		<style>
			p {
				text-decoration: none;
				color: #000;
			}
			a {
				text-decoration: none;
				color: #d14826;
			}
			a:hover,
			a:focus {
				text-decoration: underline;
			}
		</style>
	</head>
	<body>
		<p>Read about WAI on the <a href="http://w3.org/WAI">WAI webpage</a>.</p>
	</body>
</html>
```

### Failed

#### Failed Example 1

> This is a link that is a descendant of a paragraph element and the underline is removed but the link has no other way of being recognized as a link.

Reviewer 5: ```suggestion
This is a link that is a descendant of a paragraph element. The underline is removed and the link has no visual cues of being recognized as a link.
```
